File: sdks/python/kyuubiki_sdk/model_research_execution.py
from __future__ import annotations

import copy
from collections.abc import Callable, Mapping
from typing import Any, Protocol

from .errors import ModelResearchExecutionError
from .model_collaboration import MODEL_HEADLESS_PLAN_SCHEMA_VERSION
from .model_plan_approval import (
    MODEL_PLAN_APPROVAL_SCHEMA_VERSION,
    compute_model_headless_plan_digest,
)
from .session import KyuubikiSession


MODEL_RESEARCH_RECEIPT_SCHEMA_VERSION = "kyuubiki.model-research-execution-receipt/v2"
ApprovalVerifier = Callable[[Mapping[str, Any], Mapping[str, Any]], bool]


class ModelActionDispatcher(Protocol):
    def dispatch_model_action(
        self, action: str, payload: Mapping[str, Any]
    ) -> dict[str, Any]: ...


class SessionModelActionDispatcher:
    def __init__(
        self,
        session: KyuubikiSession,
        *,
        poll_interval_s: float = 0.5,
        timeout_s: float = 300.0,
    ) -> None:
        if poll_interval_s <= 0 or timeout_s <= 0 or poll_interval_s > timeout_s:
            _fail("model dispatcher wait bounds require 0 < poll_interval_s <= timeout_s")
        if timeout_s > 86_400:
            _fail("model dispatcher timeout cannot exceed 24 hours")
        self.session = session
        self.poll_interval_s = poll_interval_s
        self.timeout_s = timeout_s

    def dispatch_model_action(
        self, action: str, payload: Mapping[str, Any]
    ) -> dict[str, Any]:
        control_plane = self.session.control_plane
        if action == "direct_solver_rpc":
            output = self.session.solve_direct(
                _required_string(payload, "solve_kind"),
                _required_mapping(payload, "payload"),
            )
            return {"authority": "solver_rpc", "output": output}
        if control_plane is None:
            raise RuntimeError("control plane client is not configured")

        if action == "service_health":
            output = control_plane.health()
        elif action == "protocol_describe":
            output = control_plane.protocol()
        elif action == "agents_describe":
            output = control_plane.agents()
        elif action == "workflow_catalog_list":
            output = control_plane.list_workflow_catalog()
        elif action == "operator_catalog_list":
            output = control_plane.list_workflow_operators()
        elif action == "fem_submit":
            output = control_plane.submit_fem_job(
                _required_string(payload, "solve_kind"),
                _required_mapping(payload, "payload"),
            )
        elif action == "workflow_submit_catalog":
            output = control_plane.submit_workflow_catalog_job(
                _required_string(payload, "workflow_id"),
                _required_mapping(payload, "input_artifacts"),
            )
        elif action == "workflow_submit_graph":
            output = control_plane.submit_workflow_graph_job(
                _required_mapping(payload, "graph"),
                _required_mapping(payload, "input_artifacts"),
            )
        elif action == "operator_task_prepare":
            output = control_plane.prepare_operator_task(_required_mapping(payload, "task"))
        elif action == "operator_task_execute":
            output = control_plane.execute_operator_task(_required_mapping(payload, "task"))
        elif action == "operator_task_batch_prepare":
            output = control_plane.prepare_operator_task_batch(
                _required_mapping(payload, "batch")
            )
        elif action == "operator_task_batch_execute":
            output = control_plane.execute_operator_task_batch(
                _required_mapping(payload, "batch")
            )
        elif action == "job_wait":
            output = self.session.wait_for_job(
                _required_string(payload, "job_id"),
                poll_interval_s=self.poll_interval_s,
                timeout_s=self.timeout_s,
            )
        elif action == "result_fetch":
            output = control_plane.fetch_result(_required_string(payload, "job_id"))
        elif action == "result_chunk_fetch":
            output = control_plane.fetch_result_chunk(
                _required_string(payload, "job_id"),
                _required_string(payload, "kind"),
                offset=_optional_unsigned(payload, "offset"),
                limit=_optional_unsigned(payload, "limit"),
            )
        elif action == "job_cancel":
            output = control_plane.cancel_job(_required_string(payload, "job_id"))
        else:
            _fail(f"unsupported model action: {action}")
        return {"authority": "control_plane", "output": output}


def execute_model_headless_plan(
    dispatcher: ModelActionDispatcher,
    plan: Mapping[str, Any],
    approval: Mapping[str, Any] | None,
    approval_verifier: ApprovalVerifier,
) -> dict[str, Any]:
    plan_digest = _validate_execution_request(plan, approval)
    if approval is not None and not approval_verifier(plan, approval):
        _fail("caller approval verifier rejected approval")
    execution_plan = copy.deepcopy(plan)
    if compute_model_headless_plan_digest(execution_plan) != plan_digest:
        _fail("model plan changed after approval verification")

    records: list[dict[str, Any]] = []
    for step in execution_plan["steps"]:
        job_id = _action_job_id(step["action"], step["payload"])
        try:
            dispatched = dispatcher.dispatch_model_action(step["action"], step["payload"])
            records.append(
                {
                    "index": step["index"],
                    "action": step["action"],
                    "job_id": job_id,
                    "authority": dispatched["authority"],
                    "output": dispatched["output"],
                    "error": None,
                }
            )
        except Exception as error:
            records.append(
                {
                    "index": step["index"],
                    "action": step["action"],
                    "job_id": job_id,
                    "authority": None,
                    "output": None,
                    "error": _bounded_error(error),
                }
            )
            return _receipt(
                execution_plan, plan_digest, approval, "failed", step["index"], records
            )
    return _receipt(execution_plan, plan_digest, approval, "completed", None, records)


def _action_job_id(action: str, payload: Mapping[str, Any]) -> str | None:
    if action not in {"job_wait", "result_fetch", "result_chunk_fetch", "job_cancel"}:
        return None
    value = payload.get("job_id")
    return value if isinstance(value, str) else None


def _validate_execution_request(
    plan: Mapping[str, Any], approval: Mapping[str, Any] | None
) -> str:
    errors: list[str] = []
    if not isinstance(plan, Mapping):
        _fail("model plan must be a JSON object")
    if plan.get("schema_version") != MODEL_HEADLESS_PLAN_SCHEMA_VERSION:
        errors.append(f"unsupported model plan schema_version: {plan.get('schema_version')}")
    if not plan.get("ok") or plan.get("issues"):
        errors.append("model plan must be valid and issue-free before dispatch")
    steps = plan.get("steps")
    if not isinstance(steps, list) or not steps:
        errors.append("model plan contains no steps")
        steps = []
    if any(
        not isinstance(step, Mapping) or step.get("index") != index
        for index, step in enumerate(steps, start=1)
    ):
        errors.append("model plan step indexes must be contiguous and one-based")

    plan_digest = compute_model_headless_plan_digest(plan)
    gated = {
        (step["index"], step["action"])
        for step in steps
        if isinstance(step, Mapping) and step.get("requires_confirmation")
    }
    approved = (
        _validate_approval(plan, plan_digest, approval, gated, errors) if approval else set()
    )
    for index, action in gated - approved:
        errors.append(f"step {index} ({action}) requires an exact caller-issued approval")
    if errors:
        raise ModelResearchExecutionError(sorted(set(errors)))
    return plan_digest


def _validate_approval(
    plan: Mapping[str, Any],
    plan_digest: str,
    approval: Mapping[str, Any],
    gated: set[tuple[int, str]],
    errors: list[str],
) -> set[tuple[int, str]]:
    if not isinstance(approval, Mapping):
        errors.append("model approval must be a JSON object")
        return set()
    if approval.get("schema_version") != MODEL_PLAN_APPROVAL_SCHEMA_VERSION:
        errors.append(
            f"unsupported model approval schema_version: {approval.get('schema_version')}"
        )
    if (
        approval.get("session_id") != plan.get("session_id")
        or approval.get("workflow_id") != plan.get("workflow_id")
    ):
        errors.append("model approval does not match plan session and workflow")
    if approval.get("plan_digest") != plan_digest:
        errors.append("model approval plan_digest does not match the complete plan")
    for key in ("approval_id", "authority", "issued_at"):
        if not isinstance(approval.get(key), str) or not approval[key].strip():
            errors.append(f"model approval {key} is required")

    approved: set[tuple[int, str]] = set()
    steps = approval.get("approved_steps")
    if not isinstance(steps, list):
        errors.append("model approval approved_steps must be an array")
        return approved
    for step in steps:
        if not isinstance(step, Mapping):
            errors.append("model approval steps must be JSON objects")
            continue
        key = (step.get("index"), step.get("action"))
        if not isinstance(key[0], int) or isinstance(key[0], bool) or not isinstance(key[1], str):
            errors.append("model approval step requires integer index and string action")
            continue
        typed_key = (key[0], key[1])
        if typed_key in approved:
            errors.append(f"model approval repeats step {typed_key[0]} ({typed_key[1]})")
        approved.add(typed_key)
        if typed_key not in gated:
            errors.append(
                f"model approval references a non-gated or mismatched step {typed_key[0]} "
                f"({typed_key[1]})"
            )
    return approved


def _receipt(
    plan: Mapping[str, Any],
    plan_digest: str,
    approval: Mapping[str, Any] | None,
    status: str,
    failed_step: int | None,
    records: list[dict[str, Any]],
) -> dict[str, Any]:
    return {
        "schema_version": MODEL_RESEARCH_RECEIPT_SCHEMA_VERSION,
        "plan_schema_version": plan["schema_version"],
        "session_id": plan["session_id"],
        "workflow_id": plan["workflow_id"],
        "plan_digest": plan_digest,
        "status": status,
        "execution_authority": "kyuubiki-headless-sdk",
        "approval_id": approval.get("approval_id") if approval else None,
        "completed_steps": sum(record["error"] is None for record in records),
        "failed_step": failed_step,
        "records": records,
    }


def _required_string(payload: Mapping[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str) or not value.strip():
        _fail(f"model action payload requires non-empty string {key}")
    return value


def _required_mapping(payload: Mapping[str, Any], key: str) -> dict[str, Any]:
    value = payload.get(key)
    if not isinstance(value, Mapping):
        _fail(f"model action payload {key} must be a JSON object")
    return dict(value)


def _optional_unsigned(payload: Mapping[str, Any], key: str) -> int | None:
    value = payload.get(key)
    if value is None:
        return None
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        _fail(f"model action payload {key} must be an unsigned integer")
    return value


def _bounded_error(error: Exception) -> str:
    message = str(error)
    encoded = message.encode("utf-8")
    if len(encoded) <= 2_048:
        return message
    return f"{encoded[:2_048].decode('utf-8', errors='ignore')}..."


def _fail(message: str) -> None:
    raise ModelResearchExecutionError([message])

File: sdks/elixir/lib/kyuubiki_sdk/model_research_execution.ex
defmodule KyuubikiSdk.ModelResearchExecution do
  @moduledoc "Caller-verified execution of validated model Headless plans."

  alias KyuubikiSdk.ControlPlaneClient
  alias KyuubikiSdk.Error
  alias KyuubikiSdk.ModelCollaboration
  alias KyuubikiSdk.ModelPlanApproval
  alias KyuubikiSdk.Session

  @approval_schema "kyuubiki.model-plan-approval/v2"
  @receipt_schema "kyuubiki.model-research-execution-receipt/v2"

  def approval_schema_version, do: @approval_schema
  def receipt_schema_version, do: @receipt_schema

  def session_dispatcher(session, opts \\ []) do
    fn action, payload -> dispatch_session_action(session, action, payload, opts) end
  end

  def execute(dispatcher, plan, approval, approval_verifier)
      when is_function(dispatcher, 2) and is_function(approval_verifier, 2) do
    with {:ok, plan_digest} <- validate_execution_request(plan, approval),
         :ok <- verify_approval(plan, approval, approval_verifier),
         {:ok, verified_digest} <- ModelPlanApproval.compute_digest(plan),
         :ok <- ensure_digest_unchanged(plan_digest, verified_digest) do
      execute_steps(dispatcher, plan, plan_digest, approval)
    end
  end

  def execute(_dispatcher, _plan, _approval, _approval_verifier),
    do: validation_error("dispatcher and approval verifier must be two-argument functions")

  def dispatch_session_action(session, action, payload, opts \\ []) do
    poll_interval = Keyword.get(opts, :poll_interval, 500)
    timeout = Keyword.get(opts, :timeout, 300_000)

    with :ok <- validate_wait_bounds(poll_interval, timeout) do
      do_dispatch_session_action(session, action, payload, poll_interval, timeout)
    end
  end

  defp do_dispatch_session_action(session, "direct_solver_rpc", payload, _poll, _timeout) do
    with {:ok, solve_kind} <- required_string(payload, "solve_kind"),
         {:ok, model_payload} <- required_map(payload, "payload"),
         {:ok, output} <- Session.solve_direct(session, solve_kind, model_payload) do
      dispatch_result("solver_rpc", output)
    end
  end

  defp do_dispatch_session_action(session, action, payload, poll_interval, timeout) do
    with {:ok, client} <- control_plane(session) do
      dispatch_control_plane(session, client, action, payload, poll_interval, timeout)
    end
  end

  defp dispatch_control_plane(_session, client, "service_health", _payload, _poll, _timeout),
    do: wrap_dispatch("control_plane", ControlPlaneClient.health(client))

  defp dispatch_control_plane(_session, client, "protocol_describe", _payload, _poll, _timeout),
    do: wrap_dispatch("control_plane", ControlPlaneClient.protocol(client))

  defp dispatch_control_plane(_session, client, "agents_describe", _payload, _poll, _timeout),
    do: wrap_dispatch("control_plane", ControlPlaneClient.agents(client))

  defp dispatch_control_plane(
         _session,
         client,
         "workflow_catalog_list",
         _payload,
         _poll,
         _timeout
       ),
       do: wrap_dispatch("control_plane", ControlPlaneClient.list_workflow_catalog(client))

  defp dispatch_control_plane(
         _session,
         client,
         "operator_catalog_list",
         _payload,
         _poll,
         _timeout
       ),
       do: wrap_dispatch("control_plane", ControlPlaneClient.list_workflow_operators(client))

  defp dispatch_control_plane(_session, client, "fem_submit", payload, _poll, _timeout) do
    with {:ok, solve_kind} <- required_string(payload, "solve_kind"),
         {:ok, model_payload} <- required_map(payload, "payload") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.submit_fem_job(client, solve_kind, model_payload)
      )
    end
  end

  defp dispatch_control_plane(
         _session,
         client,
         "workflow_submit_catalog",
         payload,
         _poll,
         _timeout
       ) do
    with {:ok, workflow_id} <- required_string(payload, "workflow_id"),
         {:ok, input_artifacts} <- required_map(payload, "input_artifacts") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.submit_workflow_catalog_job(client, workflow_id, input_artifacts)
      )
    end
  end

  defp dispatch_control_plane(_session, client, "workflow_submit_graph", payload, _poll, _timeout) do
    with {:ok, graph} <- required_map(payload, "graph"),
         {:ok, input_artifacts} <- required_map(payload, "input_artifacts") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.submit_workflow_graph_job(client, graph, input_artifacts)
      )
    end
  end

  defp dispatch_control_plane(_session, client, "operator_task_prepare", payload, _poll, _timeout) do
    with {:ok, task} <- required_map(payload, "task") do
      wrap_dispatch("control_plane", ControlPlaneClient.prepare_operator_task(client, task))
    end
  end

  defp dispatch_control_plane(_session, client, "operator_task_execute", payload, _poll, _timeout) do
    with {:ok, task} <- required_map(payload, "task") do
      wrap_dispatch("control_plane", ControlPlaneClient.execute_operator_task(client, task))
    end
  end

  defp dispatch_control_plane(
         _session,
         client,
         "operator_task_batch_prepare",
         payload,
         _poll,
         _timeout
       ) do
    with {:ok, batch} <- required_map(payload, "batch") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.prepare_operator_task_batch(client, batch)
      )
    end
  end

  defp dispatch_control_plane(
         _session,
         client,
         "operator_task_batch_execute",
         payload,
         _poll,
         _timeout
       ) do
    with {:ok, batch} <- required_map(payload, "batch") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.execute_operator_task_batch(client, batch)
      )
    end
  end

  defp dispatch_control_plane(session, _client, "job_wait", payload, poll_interval, timeout) do
    with {:ok, job_id} <- required_string(payload, "job_id") do
      wrap_dispatch(
        "control_plane",
        Session.wait_for_job(session, job_id, poll_interval: poll_interval, timeout: timeout)
      )
    end
  end

  defp dispatch_control_plane(_session, client, "result_fetch", payload, _poll, _timeout) do
    with {:ok, job_id} <- required_string(payload, "job_id") do
      wrap_dispatch("control_plane", ControlPlaneClient.fetch_result(client, job_id))
    end
  end

  defp dispatch_control_plane(_session, client, "result_chunk_fetch", payload, _poll, _timeout) do
    with {:ok, job_id} <- required_string(payload, "job_id"),
         {:ok, kind} <- required_string(payload, "kind"),
         {:ok, offset} <- optional_unsigned(payload, "offset"),
         {:ok, limit} <- optional_unsigned(payload, "limit") do
      wrap_dispatch(
        "control_plane",
        ControlPlaneClient.fetch_result_chunk(client, job_id, kind,
          offset: offset,
          limit: limit
        )
      )
    end
  end

  defp dispatch_control_plane(_session, client, "job_cancel", payload, _poll, _timeout) do
    with {:ok, job_id} <- required_string(payload, "job_id") do
      wrap_dispatch("control_plane", ControlPlaneClient.cancel_job(client, job_id))
    end
  end

  defp dispatch_control_plane(_session, _client, action, _payload, _poll, _timeout),
    do: validation_error("unsupported model action: #{action}")

  defp execute_steps(dispatcher, plan, plan_digest, approval) do
    {status, failed_step, records} =
      Enum.reduce_while(plan["steps"], {:completed, nil, []}, fn step,
                                                                 {_status, _failed, records} ->
        case dispatcher.(step["action"], step["payload"]) do
          {:ok, dispatched} ->
            record = %{
              "index" => step["index"],
              "action" => step["action"],
              "job_id" => action_job_id(step["action"], step["payload"]),
              "authority" => dispatched["authority"],
              "output" => dispatched["output"],
              "error" => nil
            }

            {:cont, {:completed, nil, [record | records]}}

          {:error, reason} ->
            record = %{
              "index" => step["index"],
              "action" => step["action"],
              "job_id" => action_job_id(step["action"], step["payload"]),
              "authority" => nil,
              "output" => nil,
              "error" => bounded_error(reason)
            }

            {:halt, {:failed, step["index"], [record | records]}}

          other ->
            record = %{
              "index" => step["index"],
              "action" => step["action"],
              "job_id" => action_job_id(step["action"], step["payload"]),
              "authority" => nil,
              "output" => nil,
              "error" => bounded_error({:invalid_dispatch_result, other})
            }

            {:halt, {:failed, step["index"], [record | records]}}
        end
      end)

    records = Enum.reverse(records)

    {:ok,
     %{
       "schema_version" => @receipt_schema,
       "plan_schema_version" => plan["schema_version"],
       "session_id" => plan["session_id"],
       "workflow_id" => plan["workflow_id"],
       "plan_digest" => plan_digest,
       "status" => Atom.to_string(status),
       "execution_authority" => "kyuubiki-headless-sdk",
       "approval_id" => if(approval, do: approval["approval_id"], else: nil),
       "completed_steps" => Enum.count(records, &is_nil(&1["error"])),
       "failed_step" => failed_step,
       "records" => records
     }}
  end

  defp action_job_id(action, payload)
       when action in ["job_wait", "result_fetch", "result_chunk_fetch", "job_cancel"] and
              is_map(payload),
       do: payload["job_id"]

  defp action_job_id(_action, _payload), do: nil

  defp validate_execution_request(plan, approval) when is_map(plan) do
    steps = if is_list(plan["steps"]), do: plan["steps"], else: []

    errors =
      []
      |> add_issue(
        plan["schema_version"] != ModelCollaboration.plan_schema_version(),
        "unsupported model plan schema_version: #{inspect(plan["schema_version"])}"
      )
      |> add_issue(
        plan["ok"] != true or plan["issues"] not in [nil, []],
        "model plan must be valid and issue-free before dispatch"
      )
      |> add_issue(steps == [], "model plan contains no steps")
      |> add_issue(
        not contiguous_steps?(steps),
        "model plan step indexes must be contiguous and one-based"
      )

    with {:ok, plan_digest} <- ModelPlanApproval.compute_digest(plan) do
      gated =
        steps
        |> Enum.filter(&(&1["requires_confirmation"] == true))
        |> MapSet.new(&{&1["index"], &1["action"]})

      {approved, errors} = validate_approval(plan, plan_digest, approval, gated, errors)

      errors =
        Enum.reduce(MapSet.difference(gated, approved), errors, fn {index, action}, current ->
          ["step #{index} (#{action}) requires an exact caller-issued approval" | current]
        end)

      if errors == [],
        do: {:ok, plan_digest},
        else: {:error, Error.model_research_execution(errors |> Enum.uniq() |> Enum.sort())}
    end
  end

  defp validate_execution_request(_plan, _approval),
    do: validation_error("model plan must be a JSON object")

  defp validate_approval(_plan, _plan_digest, nil, _gated, errors),
    do: {MapSet.new(), errors}

  defp validate_approval(plan, plan_digest, approval, gated, errors) when is_map(approval) do
    errors =
      errors
      |> add_issue(
        approval["schema_version"] != @approval_schema,
        "unsupported model approval schema_version: #{inspect(approval["schema_version"])}"
      )
      |> add_issue(
        approval["session_id"] != plan["session_id"] or
          approval["workflow_id"] != plan["workflow_id"],
        "model approval does not match plan session and workflow"
      )
      |> add_issue(
        approval["plan_digest"] != plan_digest,
        "model approval plan_digest does not match the complete plan"
      )
      |> require_approval_string(approval, "approval_id")
      |> require_approval_string(approval, "authority")
      |> require_approval_string(approval, "issued_at")

    case approval["approved_steps"] do
      steps when is_list(steps) ->
        Enum.reduce(steps, {MapSet.new(), errors}, fn step, {approved, current} ->
          validate_approved_step(step, gated, approved, current)
        end)

      _ ->
        {MapSet.new(), ["model approval approved_steps must be an array" | errors]}
    end
  end

  defp validate_approval(_plan, _plan_digest, _approval, _gated, errors),
    do: {MapSet.new(), ["model approval must be a JSON object" | errors]}

  defp validate_approved_step(step, gated, approved, errors) when is_map(step) do
    key = {step["index"], step["action"]}

    cond do
      not is_integer(elem(key, 0)) or not is_binary(elem(key, 1)) ->
        {approved, ["model approval step requires integer index and string action" | errors]}

      MapSet.member?(approved, key) ->
        {approved, ["model approval repeats step #{elem(key, 0)} (#{elem(key, 1)})" | errors]}

      not MapSet.member?(gated, key) ->
        {MapSet.put(approved, key),
         [
           "model approval references a non-gated or mismatched step #{elem(key, 0)} (#{elem(key, 1)})"
           | errors
         ]}

      true ->
        {MapSet.put(approved, key), errors}
    end
  end

  defp validate_approved_step(_step, _gated, approved, errors),
    do: {approved, ["model approval steps must be JSON objects" | errors]}

  defp verify_approval(_plan, nil, _verifier), do: :ok

  defp verify_approval(plan, approval, verifier) do
    case verifier.(plan, approval) do
      true ->
        :ok

      :ok ->
        :ok

      {:ok, _evidence} ->
        :ok

      {:error, reason} ->
        validation_error("caller approval verifier rejected approval: #{bounded_error(reason)}")

      _ ->
        validation_error("caller approval verifier rejected approval")
    end
  end

  defp ensure_digest_unchanged(digest, digest), do: :ok

  defp ensure_digest_unchanged(_expected, _actual),
    do: validation_error("model plan changed after approval verification")

  defp control_plane(%Session{control_plane: nil}),
    do: {:error, Error.transport("control plane client is not configured")}

  defp control_plane(%Session{control_plane: client}), do: {:ok, client}

  defp control_plane(_session), do: {:error, Error.transport("invalid Headless session")}

  defp wrap_dispatch(authority, {:ok, output}), do: dispatch_result(authority, output)
  defp wrap_dispatch(_authority, {:error, reason}), do: {:error, reason}

  defp dispatch_result(authority, output),
    do: {:ok, %{"authority" => authority, "output" => output}}

  defp required_string(payload, key) when is_map(payload) do
    case payload[key] do
      value when is_binary(value) and value != "" -> {:ok, value}
      _ -> validation_error("model action payload requires non-empty string #{key}")
    end
  end

  defp required_string(_payload, key),
    do: validation_error("model action payload requires non-empty string #{key}")

  defp required_map(payload, key) when is_map(payload) do
    case payload[key] do
      value when is_map(value) -> {:ok, value}
      _ -> validation_error("model action payload #{key} must be a JSON object")
    end
  end

  defp required_map(_payload, key),
    do: validation_error("model action payload #{key} must be a JSON object")

  defp optional_unsigned(payload, key) when is_map(payload) do
    case payload[key] do
      nil -> {:ok, nil}
      value when is_integer(value) and value >= 0 -> {:ok, value}
      _ -> validation_error("model action payload #{key} must be an unsigned integer")
    end
  end

  defp require_approval_string(errors, approval, key) do
    add_issue(errors, not present_string?(approval[key]), "model approval #{key} is required")
  end

  defp present_string?(value), do: is_binary(value) and String.trim(value) != ""

  defp contiguous_steps?(steps) do
    steps
    |> Enum.with_index(1)
    |> Enum.all?(fn {step, index} -> is_map(step) and step["index"] == index end)
  end

  defp validate_wait_bounds(poll_interval, timeout) do
    cond do
      not is_integer(poll_interval) or not is_integer(timeout) or poll_interval <= 0 or
        timeout <= 0 or poll_interval > timeout ->
        validation_error("model dispatcher wait bounds require 0 < poll_interval <= timeout")

      timeout > 86_400_000 ->
        validation_error("model dispatcher timeout cannot exceed 24 hours")

      true ->
        :ok
    end
  end

  defp bounded_error(%{message: message}) when is_binary(message), do: truncate_error(message)
  defp bounded_error(reason), do: reason |> inspect(limit: 20) |> truncate_error()
  defp truncate_error(message) when byte_size(message) <= 2_048, do: message
  defp truncate_error(message), do: valid_prefix(message, 2_048) <> "..."

  defp valid_prefix(message, size) do
    prefix = binary_part(message, 0, size)
    if String.valid?(prefix), do: prefix, else: valid_prefix(message, size - 1)
  end

  defp add_issue(errors, true, message), do: [message | errors]
  defp add_issue(errors, false, _message), do: errors
  defp validation_error(message), do: {:error, Error.model_research_execution([message])}
end
